File: app/src/main/java/com/maximus/productivityappfinalproject/presentation/LimitedListViewModel.java
package com.maximus.productivityappfinalproject.presentation;

import android.app.Application;
import android.content.Context;
import android.util.Log;
import android.view.View;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.arch.core.util.Function;
import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.Transformations;
import androidx.lifecycle.ViewModel;

import com.google.android.material.chip.Chip;
import com.maximus.productivityappfinalproject.R;
import com.maximus.productivityappfinalproject.data.AppLimitDataSource;
import com.maximus.productivityappfinalproject.data.AppsRepository;
import com.maximus.productivityappfinalproject.data.AppsRepositoryImpl;
import com.maximus.productivityappfinalproject.device.MyUsageStatsManagerWrapper;
import com.maximus.productivityappfinalproject.domain.LimitedListUseCase;
import com.maximus.productivityappfinalproject.domain.model.AppUsageLimitModel;
import com.maximus.productivityappfinalproject.domain.model.LimitedApps;
import com.maximus.productivityappfinalproject.utils.Utils;

import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;

import io.reactivex.Observable;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.Disposable;
import io.reactivex.schedulers.Schedulers;

public class LimitedListViewModel extends ViewModel {

    private static final String TAG = "IgnoreListViewModel";

    AppsRepository mAppsRepository;

    LimitedListUseCase mLimitedListUseCase;
    private MutableLiveData<List<LimitedApps>> mAllIgnoreItems;
    /**
     * Используем в data binding
     */

    private List<LimitedApps> mLimitedApps;
    //    private IgnoreAppDataSource mIgnoreAppDataSourceImp;
    private Context mContext;
    private MyUsageStatsManagerWrapper mMyUsageStatsManagerWrapper;
    private AppLimitDataSource mAppLimitDataSource;


    @Inject
    public LimitedListViewModel(AppsRepository appsRepository, LimitedListUseCase useCase, MyUsageStatsManagerWrapper myUsageStatsManagerWrapper) {
        this.mAppsRepository = appsRepository;
        this.mLimitedListUseCase = useCase;
        this.mMyUsageStatsManagerWrapper = myUsageStatsManagerWrapper;
        mAllIgnoreItems = new MutableLiveData<>();

    }

    public final LiveData<Boolean> empty = Transformations.map(mAllIgnoreItems,
            new Function<List<LimitedApps>, Boolean>() {
                @Override
                public Boolean apply(List<LimitedApps> input) {
                    return input.isEmpty();
                }
            });

    public LiveData<List<LimitedApps>> getAllIgnoreItems() {
        return mAllIgnoreItems;
    }

    public void deleteAll() {
        mLimitedListUseCase.deleteAllLimitedItems();
    }

    public void getLimitedItems() {
        Disposable d =  mLimitedListUseCase.getLimitedList()
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(limitedApps -> onLimitedItemsFetched(limitedApps));
    }

    private void onLimitedItemsFetched(List<LimitedApps> limitedApps) {
        mLimitedApps = new ArrayList<>(limitedApps.size());
        mAllIgnoreItems.setValue(limitedApps);
    }

    public void deleteItem(String packageName) {
        mLimitedListUseCase.deleteLimitedItem(packageName);
    }

//    public Observable<List<AppUsageLimitModel>> getLimited() {
//        return mAppsRepository.getLimitedItems();
//    }


//    public String getRemainingTime(String packageName) {
//        getLimited()
//                .flatMap((list) -> Observable.fromIterable(list))
//                .subscribe((app) -> {
//                    if (app.getPackageName().equals(packageName)) {
//                        app.getTimeLimitPerHour();
//                    }
//                }, e -> {
//                    Log.e(TAG, "getRemainingTime: ", e);
//                });
//        return null;
//    }


    public void refresh(LimitedApps limitedApps) {
        mMyUsageStatsManagerWrapper.refreshIgnoreList(limitedApps);
    }

    //TODO
    public void showLimitedTime(List<AppUsageLimitModel> list, String mPackageName, Chip mLimitHourlyChip, Chip mLimitDailyChip) {
        for (AppUsageLimitModel appUsageLimitModel : list) {
            if (appUsageLimitModel.getPackageName().equals(mPackageName)) {
                int timeLimitPerDay = appUsageLimitModel.getTimeLimitPerDay();
                int timeLimitPerHour = appUsageLimitModel.getTimeLimitPerHour();
                int timeCompleted = appUsageLimitModel.getTimeCompleted();
                Log.d(TAG, "showLimitedTime: " + appUsageLimitModel.getPackageName() + "   " + appUsageLimitModel.getTimeLimitPerHour());
                if (timeLimitPerDay >= 90000000) {
                    //TODO livedata?
                    mLimitHourlyChip.setText("Лимит не установлен");
                } else {
                    mLimitDailyChip.setText(mContext.getString(R.string.limit_in_day_chip_with_args, Utils.formatMillisToSeconds(timeLimitPerDay)));
                }
                if (timeLimitPerHour >= 90000000) {
                    mLimitHourlyChip.setText("Лимит не установлен");
                } else {
                    mLimitHourlyChip.setText(mContext.getString(R.string.hourly_limit_set_to, Utils.formatMillisToSeconds(timeLimitPerHour)));
                }

                Toast.makeText(mContext, Utils.formatMillisToSeconds(timeCompleted), Toast.LENGTH_SHORT).show();
            }
//            else {
//                mLimitDailyChip.setText("Лимит не установлен");
//                mLimitHourlyChip.setText("Лимит не установлен");
//            }
        }
    }

//    public Disposable showLimitedTimeOnChips(String packageName, Chip mLimitHourlyChip, Chip mLimitDailyChip) {
//        return getLimited()
//                .subscribeOn(Schedulers.io())
//                .observeOn(AndroidSchedulers.mainThread())
//                .subscribe(list -> {
//                    showLimitedTime(list, packageName, mLimitHourlyChip, mLimitDailyChip);
//                }, e -> {
//                    Log.d(TAG, "rxjavasomething: " + e);
//                });
//    }
}

File: app/src/main/java/com/maximus/productivityappfinalproject/presentation/viewmodels/LimitedListViewModel.kt
package com.maximus.productivityappfinalproject.presentation.viewmodels

import android.util.Log
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import com.maximus.productivityappfinalproject.data.AppsRepository
import com.maximus.productivityappfinalproject.data.AppsRepositoryImpl
import com.maximus.productivityappfinalproject.device.MyUsageStatsManagerWrapper
import com.maximus.productivityappfinalproject.domain.LimitedListUseCase
import com.maximus.productivityappfinalproject.domain.model.AppUsageLimitModel
import com.maximus.productivityappfinalproject.domain.model.LimitedApps
import com.maximus.productivityappfinalproject.domain.model.PhoneUsage
import com.maximus.productivityappfinalproject.utils.Utils
import io.reactivex.Observable
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.functions.BiFunction
import io.reactivex.schedulers.Schedulers
import java.util.*
import javax.inject.Inject

class LimitedListViewModel @Inject constructor(
        private var appsRepository: AppsRepository,
        private var limitedListUseCase: LimitedListUseCase,
        private var myUsageStatsManagerWrapper: MyUsageStatsManagerWrapper)
    : ViewModel() {

    val TAG = "LimitedListViewModel"

    @Inject
    lateinit var appRepositoryImpl: AppsRepositoryImpl
    var limitedApps = mutableListOf<LimitedApps>()
    private val mAllIgnoreItems = MutableLiveData<List<LimitedApps>>()
    val allIgnoreItems: LiveData<List<LimitedApps>>
        get() = mAllIgnoreItems;

    private val _time = MutableLiveData<String>()
    val time: LiveData<String>
        get() = _time;


    fun deleteAllLimitedItems() {
        limitedListUseCase.deleteAllLimitedItems();
    }

    fun getAllLimitedItems() {
        var d = limitedListUseCase.getLimitedList()
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe({
                    onLimitedItemsFetched(it)
                }, {
                    Log.d(TAG, "OnError")
                })
    }

    private fun onLimitedItemsFetched(allLimitedApps: List<LimitedApps>) {
        limitedApps = ArrayList(allLimitedApps)
        mAllIgnoreItems.value = limitedApps
    }

    fun getRemainTime(packageName: String) {
        var d = Observable.zip(appRepositoryImpl.phoneUsage, appRepositoryImpl.limitObservable,
                BiFunction<List<PhoneUsage>, List<AppUsageLimitModel>, Long> { t1: List<PhoneUsage>, t2: List<AppUsageLimitModel> ->
                    getRemainTimeFromDb(t1, t2, packageName)
                })
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe({
                    if (it <= 0) {
                        _time.value = "Перерасход епта"
                    } else {
                        _time.value = Utils.formatMillisToSeconds(it)
                    }
                }, {
                    Log.d(TAG, it.localizedMessage)
                })
    }

    private fun getRemainTimeFromDb(phoneUsage: List<PhoneUsage>, appUsageLimitModel: List<AppUsageLimitModel>, packageName: String): Long {
        var phoneUsageTime: Long = 0
        var appUsage: Long = 0
        var result: Long = 0
        for (phoneModel in phoneUsage) {
            if (packageName == phoneModel.packageName) {
                phoneUsageTime = phoneModel.timeCompletedInHour
            }
        }

        for (appUsageModel in appUsageLimitModel) {
            if (packageName == appUsageModel.packageName) {
                appUsage = appUsageModel.timeLimitPerHour.toLong()
            }
        }

        result = appUsage - phoneUsageTime
        Log.d(TAG, "$result")

        return result
    }

}